{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Linker Analysis of Nanopore Amplicon Sequencing Data\n",
    "\n",
    "This notebook analyzes **Nanopore long-read amplicon** sequencing datasets, which have undergone preprocessing, alignment, and frame correction.\n",
    "\n",
    "Please make sure to download the input dataset `Nanopore` from the Zenodo repository and place it in a folder named `data` at root repository level.\n",
    "\n",
    "Before using this notebook for sequence analysis, ensure that the preprocessing script `00_Nanopore_filtering_alignment_processing.sh` has been run. This script performs:\n",
    "\n",
    "1. Quality filtering of Nanopore reads (`01_Nanopore_read_filtering.py`)\n",
    "2. Alignment to the reference sequence (`02_Nanopore_alignment.py`)\n",
    "3. Correction of indel errors typical of Nanopore sequencing, including frame enforcement to enable codon-level mutation analysis (`03_Nanopore_quality_control.py`)\n",
    "\n",
    "Alternatively, you may run the scripts in the brackets in the given order to achieve the same result.\n",
    "\n",
    "For more details, see the `README.md` file in this repository.\n",
    "\n",
    "\n",
    "---\n",
    "\n",
    "#### This notebook includes the following steps:\n",
    "- Calculation of mutation enrichment based on aligned Nanopore reads\n",
    "- Extraction of enrichment data within the defined AraC-LOV2 region of interest\n",
    "- Quantification and visualization of insertion and deletion (indel) frequencies across the reference sequence\n",
    "- Export of enrichment metrics (absolute, relative, and variant-level) for downstream analysis\n",
    "---\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Notebook Setup\n",
    "----------------------------\n",
    "\n",
    "Run the following cell to import all required libraries, define plotting settings, and configure the environment for data analysis and visualization."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/alicehovsepyan/opt/miniconda3/lib/python3.9/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.0\n",
      "  warnings.warn(f\"A NumPy version >={np_minversion} and <{np_maxversion}\"\n"
     ]
    }
   ],
   "source": [
    "# --- Import necessary modules ---\n",
    "import os\n",
    "import sys\n",
    "import json\n",
    "import ast\n",
    "import csv\n",
    "import gzip\n",
    "import pickle as pkl\n",
    "from importlib import reload\n",
    "from pathlib import Path\n",
    "\n",
    "notebook_dir = Path().resolve()\n",
    "repo_root = notebook_dir.parent.parent \n",
    "sys.path.append(str(repo_root))\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "import scipy\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqIO import QualityIO\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import seaborn as sns\n",
    "from matplotlib import patches\n",
    "from matplotlib import cm\n",
    "from matplotlib import colors as mcolors\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib.font_manager import FontProperties\n",
    "from matplotlib.gridspec import GridSpec\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "import seaborn as sns\n",
    "import pysam\n",
    "from scripts.utils import dna_rev_comp, translate_dna2aa\n",
    "from scripts.linker_analysis_functions import *\n",
    "from scripts.preprocessing_functions import *\n",
    "from scripts.plotting import *\n",
    "from scripts.functions_ import *\n",
    "from scripts.Nanopore_functions import *\n",
    "\n",
    "# ======================== PLOTTING SETTINGS ========================\n",
    "\n",
    "# --- Custom color map for mutation visualization ---\n",
    "custom_cmap = LinearSegmentedColormap.from_list(\"custom_cmap\", [\n",
    "    \"#22577A\",  # Deep blue\n",
    "    \"#38A3A5\",  # Teal\n",
    "    \"#57CC99\",  # Medium green\n",
    "    \"#80ED99\",  # Bright green\n",
    "    \"#C7F9CC\"   # Light pastel green\n",
    "], N=256)\n",
    "\n",
    "# --- Seaborn theme configuration ---\n",
    "custom_params = {\n",
    "    \"axes.spines.right\": False,\n",
    "    \"axes.spines.top\": False,\n",
    "    \"axes.linewidth\": 1\n",
    "}\n",
    "sns.set_theme(context=\"paper\", style='ticks', palette=\"Greys_r\", rc=custom_params)\n",
    "\n",
    "# --- General matplotlib settings ---\n",
    "fs = 8  # font size\n",
    "plt.rcParams['svg.fonttype'] = 'none'\n",
    "mpl.rcParams.update({\n",
    "    'font.family': 'Avenir Next',\n",
    "    'font.weight': 'demi', \n",
    "    'font.size': fs,\n",
    "    'text.color': '#231F20',\n",
    "    'axes.labelcolor': '#231F20',\n",
    "    'xtick.color': '#231F20',\n",
    "    'ytick.color': '#231F20',\n",
    "    'axes.edgecolor': '#231F20',\n",
    "    'figure.facecolor': 'white',\n",
    "    'axes.facecolor': 'white',\n",
    "    'pdf.fonttype': 42,\n",
    "    'text.usetex': False\n",
    "})\n",
    "sns.set_context(\"paper\", rc={\n",
    "    \"font.size\": fs,\n",
    "    \"axes.titlesize\": fs + 1,\n",
    "    \"axes.labelsize\": fs,\n",
    "    \"axes.linewidth\": 1,\n",
    "    \"xtick.labelsize\": fs,\n",
    "    \"ytick.labelsize\": fs,\n",
    "    \"legend.fontsize\": fs,\n",
    "    \"legend.title_fontsize\": fs + 1\n",
    "})\n",
    "\n",
    "avenir_fp = FontProperties(family='Avenir Next', weight='demi', size=fs)\n",
    "\n",
    "# ======================== LAYOUT PARAMETERS ========================\n",
    "\n",
    "# --- Pre-defined figure dimensions for multi-panel plots ---\n",
    "hi = 10.5 * 0.75\n",
    "wi3 = 2.24\n",
    "wi2 = 4.76\n",
    "wi1 = 7.24\n",
    "nr = 5\n",
    "\n",
    "# --- Utility for shortening sample names ---\n",
    "short_fn = np.vectorize(lambda x: x[:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ======================== ANALYSIS PARAMETERS ========================\n",
    "\n",
    "ref_R2 = \"taatggaaacttcctcatgaaaaagtctttagtcctcaaagcctctgtagccgttgctaccctcgttccgatgctgtctttcgctgctgagggtgacgatcccgcaaaagcggcctttgactccctgcaagcctcagcgaccgaatatatcggttatgcgtgggcgatggttgttgtcattgtcggcgcaactatcggtatcaagctgtttaagaaattcacctcgaaagcaagttgataaactgatacaattaaaggctccttttggagcctttttttttggagtaaggaggaaaaAtgtccgcgAaagcgcagaacgatccgctgctgccgggctatagctttaacgcAcatctggtggcgggcctgaccccgattgaagcgaacggctatctggatttttttattgatcgcccgctgggcatgaaaggctatattctgaacctgaccattcgcggccagggcgtggtgaaaaaccagggccgcgaatttgtgtgccgcccgggcgatattctgctgtttccgccgggcgaaattcatcattatggccgccatccggaagcgcgAgaatggtatcatcagtgggtgtattttcgcccgcgcgcgtattggcatgaatggctgaactggccgagcatttttgcgaacaccggcttttttcgcccggatgaagcgcatcagccgcattttagcgatctgtttggccagattattaacgcgggccagggcgaaggccTctatagcgaactgctgAcAattaacctgctggaacagctgctgctgcgccgCATGGAAGCGATTAACGAAAGCagcggtttagccacaacgctggaacgcattgaaaagaatttcgtaatcacagacccgcgccttcccgacaatccaattatttttgcgtccgatagcttcctgcaattaaccgaatacagccgcgaagaaattctgggtcgtaattgtcgcttccttcaggggccagagactgaccgtgctacggtacgcaaaatccgcgacgcaatcgacaatcaaacggaagtcacggttcagttgattaactatacgaagagcggaaaaaaattctggaatttatttcacttgcagcctatgcgtgaccagaagggcgatgtccagtatttcattggcgttcagcttgatggtaccgagcatgttcgcgatgctgcggagcgtgaaggtgtaatgttaattaaaaagactgctgaaaacattgatgaggcggccaaagggagcCTGCATCCGCCGATGGATAACCGcgtgcgcgaagcgtgccagtatattagcgatcatctggcggatagcaactttgatattgcgagcgtggcgcagcatgtgtgcctgagcccTagccgcctgagccatctgtttcgccagcagctgggcattagcgtgctgagctggcgcgaagatcagcgcattagccaggcgaaactgctgctgagcaccacccgcatgccgattgcgaccgtgggccTcaacgtgggctttgatgatcagctgtattttagccgcgtgtttaaaaaatgcaccggcgcgagcccgagcgaatttcgcgcgggctgcgaagaaaaagtGAACGATGTGGCGGTGAAACTGAGCGGGTAAggctaatggagattttcaacatgggctagcacagccctaggtattatgctagcgtggtgtctgcgtaataaggagtcttaatcatgccagttcttttgggtattccgttattattgcgtttcctcggtttccttctggtaactttgttcggctatctgcttacttttctcaaaaagggcttcggtaagatagctattgctatttcattgtttcttgctcttattattgggcttaactcaattcttgtgggttatctctctgatattagtgctcaattaccctctgactttgttcagggtgttcagttaattctcccgtctaatgcgcttccctgtttttatgttattctctctgt\".upper() ## R2\n",
    "\n",
    "cut_site_seq_left= \"gccacaa\".upper()\n",
    "cut_site_seq_right= \"tgctgaaaac\".upper()\n",
    "\n",
    "wt_linker_left = \"INESSGL\" \n",
    "wt_linker_right = \"IDEAAKGSHLPP\"\n",
    "\n",
    "quality = 20\n",
    "data_folder = 'Nanopore/Nanopore_P0115'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Analysis and Renaming of Linker Variants Across Barcodes\n",
    "\n",
    "In this step, we analyze the distribution and co-occurrence of linker variants across all barcodes, including standard bar plots for visualizing variant frequencies.\n",
    "Linker names are renamed using the `linker_renaming.txt` file. Please verify that the renaming was applied correctly, as rare or unexpected mutations may interfere with the renaming process."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# --- Loop over barcodes ---\n",
    "barcodes = ['R2-LOV_RAMPhaGE_Linker_Mut_D1', 'R2-LOV_P1_Neg_D2', 'R2-LOV_P2_Neg_D2', 'R2-LOV_P1_Pos_D3', 'R2-LOV_P2_Pos_D3', 'R2-LOV_P1_Pos_D4', 'R2-LOV_P2_Pos_D\\4']\n",
    "for barcode in barcodes:\n",
    "    # --- Define input and output folders ---\n",
    "    input_folder = f\"{repo_root}/data/{data_folder}/{barcode}/highly_accurate_basecalling/filtered_Q{quality}_maxminlen/\"\n",
    "    FigFolder = f\"{repo_root}/final_output/{data_folder}/{barcode}/\"\n",
    "    if not os.path.exists(FigFolder):\n",
    "        os.makedirs(FigFolder)\n",
    "\n",
    "    # --- Extract left and right linker regions from aligned reads ---\n",
    "    all_left_linkers, all_right_linkers = get_linker_regions(input_folder= input_folder+ \"minimap2_alignment\", ref = ref_R2, cut_site_seq_left=cut_site_seq_left, cut_site_seq_right= cut_site_seq_right, left_linker_region_len= 12*3, right_linker_region_len= 20*3)\n",
    "\n",
    "\n",
    "    ######### LEFT LINKER ANALYSIS #########\n",
    "    read_dir = \"R1\"\n",
    "    # --- Count linker variants and extract list of raw alignments ---\n",
    "    left_linkers_counts, left_linker_list = get_linker_variants(linker_alignments=all_left_linkers,  wt_linker = wt_linker_left if read_dir==\"R1\" else wt_linker_right, read_dir = read_dir)\n",
    "\n",
    "    # --- Sort variants by frequency ---\n",
    "    linkers_sorted = {k: v for k, v in sorted(left_linkers_counts.items(), key=lambda item: item[1], reverse=True)}\n",
    "    total_reads = sum(linkers_sorted.values())\n",
    "    linkers_sorted_perc = {k: v/total_reads*100 for k, v in linkers_sorted.items()}\n",
    "\n",
    "    # --- Remove wild-type and variants with less than 10% frequency ---\n",
    "    linkers_sorted_perc.pop(\"wt\")\n",
    "    linkers_perc_filt, left_linker_renaming = rename_left_linkers(linkers_sorted_perc.keys(), linkers_sorted_perc)\n",
    "    linkers_perc_filt = {k: v for k, v in linkers_perc_filt.items() if v > 0.1}\n",
    "\n",
    "\n",
    "    # --- Plot left linker variant distribution ---\n",
    "    fig, ax = plt.subplots(1, 1, figsize=(15, 5))\n",
    "    plt.bar(linkers_perc_filt.keys(), linkers_perc_filt.values(), color =\"#22577A\")\n",
    "    plt.xticks(rotation=90)\n",
    "    plt.ylabel(\"Percentage of reads\")\n",
    "    plt.text(0.9, 0.93, f\"Total mutation rate: {round(sum(linkers_perc_filt.values()),3)}\", horizontalalignment='center', verticalalignment='center', transform=ax.transAxes)\n",
    "    plt.title(f\"Linker variants for {barcode} {read_dir}\")\n",
    "    # plt.savefig(f\"{FigFolder}/linker_distribution_{read_dir}.pdf\", bbox_inches=\"tight\")\n",
    "    # plt.savefig(f\"{FigFolder}/linker_distribution_{read_dir}.png\", bbox_inches=\"tight\")\n",
    "    plt.close()\n",
    "\n",
    "    # --- Save left linker variant frequencies to CSV ---\n",
    "    linkers_perc_filt = pd.DataFrame.from_dict(linkers_perc_filt, orient=\"index\")\n",
    "    linkers_perc_filt.to_csv(f\"{FigFolder}/linker_distribution_{read_dir}.csv\")\n",
    "\n",
    "\n",
    "    ######### RIGHT LINKER ANALYSIS #########\n",
    "    read_dir = \"R2\"\n",
    "    # --- Count right linker variants and extract list ---\n",
    "    right_linkers_counts, right_linkers_list = get_linker_variants(linker_alignments=all_right_linkers, wt_linker = wt_linker_left if read_dir==\"R1\" else wt_linker_right, read_dir = read_dir)\n",
    "\n",
    "    # --- Sort and filter right linker variants ---\n",
    "    linkers_sorted = {k: v for k, v in sorted(right_linkers_counts.items(), key=lambda item: item[1], reverse=True)}\n",
    "    total_reads = sum(linkers_sorted.values())\n",
    "    linkers_sorted_perc = {k: v/total_reads*100 for k, v in linkers_sorted.items()}\n",
    "    # --- Remove wild-type and variants with less than 10% frequency ---\n",
    "    linkers_sorted_perc.pop(\"wt\")\n",
    "    linkers_perc_filt, right_linker_renaming = rename_right_linkers(linkers_sorted_perc.keys(), linkers_sorted_perc)\n",
    "    linkers_perc_filt = {k: v for k, v in linkers_perc_filt.items() if v > 0.1}\n",
    "\n",
    "    # --- Plot right linker variant distribution ---\n",
    "    fig, ax = plt.subplots(1, 1, figsize=(15, 5))\n",
    "    plt.bar(linkers_perc_filt.keys(), linkers_perc_filt.values(), color =\"#22577A\")\n",
    "    plt.xticks(rotation=90)\n",
    "    plt.ylabel(\"Percentage of reads\")\n",
    "    plt.text(0.9, 0.93, f\"Total mutation rate: {round(sum(linkers_perc_filt.values()),3)}\", horizontalalignment='center', verticalalignment='center', transform=ax.transAxes)\n",
    "    plt.title(f\"Linker variants for {barcode} {read_dir}\")\n",
    "    # plt.savefig(f\"{FigFolder}/linker_distribution_{read_dir}.pdf\", bbox_inches=\"tight\")\n",
    "    # plt.savefig(f\"{FigFolder}/linker_distribution_{read_dir}.png\", bbox_inches=\"tight\")\n",
    "    plt.close()\n",
    "\n",
    "    # --- Save right linker variant frequencies to CSV ---\n",
    "    linkers_perc_filt = pd.DataFrame.from_dict(linkers_perc_filt, orient=\"index\")\n",
    "    linkers_perc_filt.to_csv(f\"{FigFolder}/linker_distribution_{read_dir}.csv\")\n",
    "\n",
    "\n",
    "    ######### COOCCURRENCE ANALYSIS #########\n",
    "    cooccurrences = []\n",
    "    for idx, right_linker in enumerate(right_linkers_list): \n",
    "        if right_linker == \"\": \n",
    "            continue\n",
    "        left_linker = left_linker_list[idx] \n",
    "        right_linker_renamed = right_linker_renaming[right_linker] if right_linker != \"wt\" else \"wt\"\n",
    "        left_linker_renamed = left_linker_renaming[left_linker] if left_linker != \"wt\" else \"wt\"\n",
    "\n",
    "        cooccurrences.append(f\"{left_linker_renamed}_{right_linker_renamed}\")\n",
    "    \n",
    "    # --- Save cooccurrence combinations to file ---\n",
    "    with open(f\"{FigFolder}/cooccurrences.txt\", \"w\") as f: \n",
    "        f.write(\"\\n\".join(cooccurrences))\n",
    "    \n",
    "    # --- Save linker renaming dictionaries as JSON ---\n",
    "    with open(f\"{FigFolder}/linker_renaming_left_linker.json\", \"w\") as f:\n",
    "        json.dump(left_linker_renaming, f, indent=4)\n",
    "    \n",
    "    with open(f\"{FigFolder}/linker_renaming_right_linker.json\", \"w\") as f:\n",
    "        json.dump(right_linker_renaming, f, indent=4)\n",
    "    \n",
    "    # --- Save raw linker alignment data to JSON ---\n",
    "    if not os.path.exists(f\"{input_folder}/processed_reads\"):\n",
    "        os.makedirs(f\"{input_folder}/processed_reads\")\n",
    "        \n",
    "    with open(f\"{input_folder}/processed_reads/all_left_linkers.json\", \"w\") as f:\n",
    "        json.dump(all_left_linkers, f, indent=4)\n",
    "    \n",
    "    with open(f\"{input_folder}/processed_reads/all_right_linkers.json\", \"w\") as f:\n",
    "        json.dump(all_right_linkers, f, indent=4)\n",
    "        \n",
    "    print(\"Barcode\", barcode, \"done\")   \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
