{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "9TYLAPpgppBl"
      },
      "source": [
        "# Oligo Tool"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "kJzwx09QG24Q"
      },
      "source": [
        "## How it works:\n",
        "\n",
        "Retrons are genetically encoded RNA/ssDNA sequences targetable to edit specified regions of homologous DNA when their reverse transcriptase and a set of recombineering proteins are expressed. This allows you to  to program mutations, insertions, and deletions at your site of interest.\n",
        "\n",
        "\n",
        "This is a tool that allows you to generate libraries of oligos that can be inserted into a retron vector via golden gate cloning. Although this was made with a specific application (Phage Assisted Evolution) and vector (RcP9) in mind to program mutations on M13 Phage, this hypothetically could be extended to generating retron libraries of any sort.\n",
        "\n",
        "\n",
        "## Usage:\n",
        "\n",
        "The tool is split into three main components:  \n",
        "\n",
        "### Barcode Generator:\n",
        "- **barcode_number**: Fetches a bioorthogonal barcode set (doi.org/10.1073/pnas.0812506106) from 3000 pairs, according to the input number. This allows the user to tag a generated library set. Forward and Reverse Primers for library PCR amplification of that pair are automatically created as downloadable fasta files in the file folder.\n",
        "\n",
        "### Systematic Screen\n",
        "The systematic screen enables the generation of a library with a predefined set of changes spread across a region of interest. Changes can include indels.\n",
        "\n",
        "**Parameters:**\n",
        "- **experiment_name**: Name of the experiment. Determines also the name of the file that is saved.  \n",
        "- **sequence**: Target sequence in which changes are introduced.\n",
        "  \n",
        "  *Note: Extra bases need to be added to the beginning and end of the target to maintain retron homology arms\n",
        "  (IE: Retron of 90bp likely requires 45bp addition at the beginning to target the first base, or else the retron would be unable to bind to a homologous sequence of interest).*\n",
        "- **barcode**: Sequences added to the beginning and end of the oligo for library tagging and final PCR amplification.\n",
        "- **goldengate_site**: Recognition Sequence of the Type IIS restriction enzyme for golden gate cloning.\n",
        "- **overhang**: Golden Gate Cloning Sticky ends generated by the Type IIS enzyme.\n",
        "- **start_position**: Start position at which the screen will start.\n",
        "- **window_size**: Determines the nucleotides that are replaced with sequences from the pool.\n",
        "If the window size is equivalent to the extension size, a mutation occurs.\n",
        "A larger window size than the extension size would lead to deletions and/or mutations depending on the extension.\n",
        "A smaller window size would lead to insertions and/or mutations depending on the extension.\n",
        "- **step_size**: Defines how much the window is moved before the extensions are applied again. Defaults to three to match a codon.\n",
        "- **end_position**: End position at which the screen will end.\n",
        "- **retron_size**: Target size of the final retron.\n",
        "- **extension**: Defines the pool of bases that are systematically inserted in the specified window.\n",
        "\n",
        "\n",
        "### Linker Screen\n",
        "The linker screen enables the user to rapidly generate constructs with varying linker types defined in the linker pool at a given position.\n",
        "\n",
        "**Parameters:**\n",
        "- **experiment_name**: Name of the experiment. Determines also the name of the file that is saved.  \n",
        "- **sequence**: Target nucleotide sequence in which changes are introduced.\n",
        "- **barcode**: Nucleotide sequences added to the beginning and end of the oligo for library tagging and final PCR amplification.\n",
        "- **goldengate_site**: Recognition Sequence of the Type IIS restriction enzyme for golden gate cloning.\n",
        "- **overhang**: Golden Gate Cloning Sticky ends generated by the Type IIS enzyme.\n",
        "- **target_position**: Position at which the linker pool will be applied.\n",
        "- **retron_size**: Target size of the final retron.\n",
        "- **linker_pool**: Amino acid sequence of comma seperated linker sequences. '-' and '+' mark deletions up or downstream of the target site. A 'X' will generate every possible amino acid at a given position.\n",
        "- **species**: Automatically adjust codons to be optimal for e.coli\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "o8n2urO7pZim"
      },
      "outputs": [],
      "source": [
        "# @title\n",
        "import itertools as it\n",
        "from itertools import product\n",
        "import re\n",
        "import os\n",
        "\n",
        "class bcolors:\n",
        "    HEADER = '\\033[95m'\n",
        "    OKBLUE = '\\033[94m'\n",
        "    OKCYAN = '\\033[96m'\n",
        "    OKGREEN = '\\033[92m'\n",
        "    WARNING = '\\033[93m'\n",
        "    FAIL = '\\033[91m'\n",
        "    ENDC = '\\033[0m'\n",
        "    BOLD = '\\033[1m'\n",
        "    UNDERLINE = '\\033[4m'\n",
        "\n",
        "\n",
        "\n",
        "codon_to_aa = {'TTT':'F','TTC':'F','TTA':'L','TTG':'L',\n",
        "              'CTT':'L','CTC':'L','CTA':'L','CTG':'L',\n",
        "              'ATT':'I','ATC':'I','ATA':'I','ATG':'M',\n",
        "              'GTT':'V','GTC':'V','GTA':'V','GTG':'V',\n",
        "\n",
        "               'TCT':'S','TCC':'S','TCA':'S','TCG':'S',\n",
        "              'CCT':'P','CCC':'P','CCA':'P','CCG':'P',\n",
        "              'ACT':'T','ACC':'T','ACA':'T','ACG':'T',\n",
        "              'GCT':'A','GCC':'A','GCA':'A','GCG':'A',\n",
        "\n",
        "               'TAT':'Y','TAC':'Y','TAA':'\\n STOP \\n','TAG':'\\n STOP \\n',\n",
        "              'CAT':'H','CAC':'H','CAA':'Q','CAG':'Q',\n",
        "              'AAT':'N','AAC':'N','AAA':'K','AAG':'K',\n",
        "              'GAT':'D','GAC':'D','GAA':'E','GAG':'E',\n",
        "\n",
        "               'TGT':'C','TGC':'C','TGA':'\\n STOP \\n','TGG':'W',\n",
        "               'CGT':'R','CGC':'R','CGA':'R','CGG':'R',\n",
        "               'AGT':'S','AGC':'S','AGA':'R','AGG':'R',\n",
        "               'GGT':'G','GGC':'G','GGA':'G','GGG':'G',\n",
        "\n",
        "              }\n",
        "\n",
        "invert_base_seq = {'A':'T','T':'A','G':'C','C':'G'}\n",
        "\n",
        "base_mapping = {'R': ['G','A'],\n",
        "                'Y': ['C','T'],\n",
        "                'S': ['G','C'],\n",
        "                'W': ['A','T'],\n",
        "                'K': ['G','T'],\n",
        "                'M': ['C','A'],\n",
        "                'A':['A'],\n",
        "                'G':['G'],\n",
        "                'C':['C'],\n",
        "                'T':['T'],\n",
        "                'U':['T'],\n",
        "                'N':['A','T','C','G']}\n",
        "\n",
        "ecoli_pref = {\n",
        "            \"A\": 'GCG',\n",
        "            \"R\": 'CGT',\n",
        "            \"N\": 'AAC',\n",
        "            \"D\": 'GAT',\n",
        "            \"C\": 'TGC',\n",
        "            \"Q\": 'CAG',\n",
        "            \"E\": 'GAA',\n",
        "            \"G\": 'GGC',\n",
        "            \"H\": 'CAT',\n",
        "            \"I\": 'ATT',\n",
        "            \"L\": \"CTG\",\n",
        "            \"K\": 'AAA',\n",
        "            \"M\": 'ATG',\n",
        "            \"F\": \"TTT\",\n",
        "            \"P\": 'CCG',\n",
        "            \"S\": 'AGC',\n",
        "            \"T\": 'ACC',\n",
        "            \"W\": 'TGG',\n",
        "            \"Y\": \"TAT\",\n",
        "            \"V\": 'GTG',\n",
        "}\n",
        "ecoli_pref_hydrophilic = {\n",
        "#             \"A\": 'GCG',\n",
        "            \"R\": 'CGT',\n",
        "            \"N\": 'AAC',\n",
        "            \"D\": 'GAT',\n",
        "            \"C\": 'TGC',\n",
        "            \"Q\": 'CAG',\n",
        "            \"E\": 'GAA',\n",
        "            \"G\": 'GGC',\n",
        "            \"H\": 'CAT',\n",
        "#             \"I\": 'ATT',\n",
        "#             \"L\": \"CTG\",\n",
        "            \"K\": 'AAA',\n",
        "#             \"M\": 'ATG',\n",
        "#             \"F\": \"TTT\",\n",
        "            \"P\": 'CCG',\n",
        "            \"S\": 'AGC',\n",
        "            \"T\": 'ACC',\n",
        "#             \"W\": 'TGG',\n",
        "#             \"Y\": \"TAT\",\n",
        "#             \"V\": 'GTG',\n",
        "}\n",
        "ecoli_pref_hydrophobic = {\n",
        "            \"A\": 'GCG',\n",
        "#             \"R\": 'CGT',\n",
        "#             \"N\": 'AAC',\n",
        "#             \"D\": 'GAT',\n",
        "#             \"C\": 'TGC',\n",
        "#             \"Q\": 'CAG',\n",
        "#             \"E\": 'GAA',\n",
        "#             \"G\": 'GGC',\n",
        "#             \"H\": 'CAT',\n",
        "            \"I\": 'ATT',\n",
        "            \"L\": \"CTG\",\n",
        "#             \"K\": 'AAA',\n",
        "            \"M\": 'ATG',\n",
        "            \"F\": \"TTT\",\n",
        "#             \"P\": 'CCG',\n",
        "#             \"S\": 'AGC',\n",
        "#             \"T\": 'ACC',\n",
        "            \"W\": 'TGG',\n",
        "            \"Y\": \"TAT\",\n",
        "            \"V\": 'GTG',\n",
        "}\n",
        "conv_type_dict = {'ecoli':ecoli_pref,\n",
        "                 'ecoli_hydrophilic':ecoli_pref_hydrophilic,\n",
        "                 'ecoli_hydrophobic':ecoli_pref_hydrophobic}\n",
        "\n",
        "special_token_mapping = {'ecoli': list(ecoli_pref.values()),\n",
        "                        'ecoli_hydrophilic': list(ecoli_pref_hydrophilic.values()),\n",
        "                        'ecoli_hydrophobic': list(ecoli_pref_hydrophobic.values())}\n",
        "aa_to_codon = {}\n",
        "\n",
        "for key, value in codon_to_aa.items():\n",
        "    if aa_to_codon.get(value) is None:\n",
        "        aa_to_codon[value] = [key]\n",
        "    else:\n",
        "        aa_to_codon[value].append(key)\n",
        "\n",
        "\n",
        "def aa_to_codon_convert(sequence,conv_type):\n",
        "\n",
        "    codon_list = []\n",
        "    for letter in sequence:\n",
        "        if conv_type_dict[conv_type].get(letter.upper()) is None:\n",
        "            return None\n",
        "        else:\n",
        "            codon_list.append(conv_type_dict[conv_type][letter.upper()])\n",
        "    codons = ''.join(codon_list)\n",
        "\n",
        "    return codons\n",
        "\n",
        "\n",
        "\n",
        "def convert_to_aa(codon_list:list):\n",
        "\n",
        "    aa_list =[codon_to_aa[codon] for codon in codon_list]\n",
        "    return aa_list\n",
        "def gen_fasta(overall_pool):\n",
        "    for names, sequences in overall_pool.items():\n",
        "        yield (f'>{names}\\n{sequences}\\n')\n",
        "\n",
        "def generate_X_variants(sequence,species):\n",
        "    aa_buffer = [[]]\n",
        "    for aa in sequence:\n",
        "\n",
        "        if aa == 'X':\n",
        "            new_buffer = []\n",
        "            for cur_seq in aa_buffer:\n",
        "                # start_seq = ''.join(cur_seq)\n",
        "                aa_add = []\n",
        "                for aa2 in 'ARNDCQEGHILKMFPSTWYV':\n",
        "                    if aa_to_codon_convert(aa2,species) is not None:\n",
        "                        aa_add.append([*cur_seq,aa_to_codon_convert(aa2,species)] )\n",
        "\n",
        "                new_buffer.extend(aa_add)\n",
        "            aa_buffer = new_buffer\n",
        "        else:\n",
        "            for cur_seq in (aa_buffer):\n",
        "                cur_seq.append(aa_to_codon_convert(aa,conv_type='ecoli'))\n",
        "    aa_buffer = [''.join(seq) for seq in aa_buffer]\n",
        "    return aa_buffer\n",
        "\n",
        "def generate_retrons(sequence: str,target_position: list,extension: str = None, predefined_pool: list = None,retron_size = 90, #bp\n",
        "                     barcodes = ('',''),overhangs=('ATTC','GTAC'), goldengate_sites = ('CGTCTCA','TGAGACG'),):\n",
        "    '''\n",
        "\n",
        "    :param sequence:\n",
        "    :param extension:\n",
        "    :param target_position:\n",
        "    :param retron_size:\n",
        "    :param barcodes:\n",
        "    :param goldengate_sites:\n",
        "    :param overhangs:\n",
        "    :return:\n",
        "    '''\n",
        "\n",
        "\n",
        "\n",
        "    retron_pool = []\n",
        "\n",
        "    bl,br = barcodes\n",
        "    ovl,ovr = overhangs\n",
        "    goldl,goldr = goldengate_sites\n",
        "\n",
        "\n",
        "\n",
        "    if extension is not None:\n",
        "        seqs = []\n",
        "        for part in extension.split(':'):\n",
        "            if special_token_mapping.get(part,None) is None:\n",
        "                seqs.extend([base_mapping[base] for base in part])\n",
        "            else:\n",
        "                seqs.append(special_token_mapping[part])\n",
        "        combinations = list(product(*seqs))\n",
        "    elif combinations is not None:\n",
        "        combinations = predefined_pool\n",
        "\n",
        "    else:\n",
        "        raise ValueError('Need a extension or target pool')\n",
        "\n",
        "    insert_pool = [''.join(combo) for combo in combinations]\n",
        "\n",
        "\n",
        "    size = retron_size-len(insert_pool[0])\n",
        "    retron_size_left,retron_size_right = (size//2,size//2) if size%2 == 0 else (size//2,(size//2)+1)\n",
        "\n",
        "\n",
        "    if isinstance(target_position,int):\n",
        "        print(f\"\\n{sequence[target_position-retron_size_left:target_position]}{bcolors.FAIL}{bcolors.BOLD}!INSERT!{bcolors.ENDC}{sequence[target_position:target_position+retron_size_right]}\")\n",
        "        for insert in insert_pool:\n",
        "            retron = f\"{sequence[target_position-retron_size_left:target_position]}|{insert}|{sequence[target_position:target_position+retron_size_right]}\"\n",
        "            assert len(retron)-2 == retron_size\n",
        "            # print(f\"{sequence[target_position-retron_size_left:target_position]}{bcolors.OKGREEN}{insert}{bcolors.ENDC}{sequence[target_position:target_position+retron_size_right]}\")\n",
        "\n",
        "\n",
        "\n",
        "            fwd_primer = f'{bl}{goldl}{ovl}'\n",
        "            rev_primer = ''.join([invert_base_seq[nt] for nt in f'{ovr}{goldr}{br}'[::-1]])\n",
        "\n",
        "\n",
        "            retron_pool.append((insert,\n",
        "                                fwd_primer,\n",
        "                               rev_primer,\n",
        "                               f'{bl}|{goldl}|{ovl}|{retron}|{ovr}|{goldr}|{br}'))\n",
        "    else:\n",
        "        start,end = target_position\n",
        "        if (end-start)%3 != 0:\n",
        "            print(f'{bcolors.WARNING}[WARNING] - The current selection might lead to a frameshift!{bcolors.ENDC}')\n",
        "\n",
        "\n",
        "        print(f\"{sequence[start-retron_size_left:start]}{bcolors.FAIL}{bcolors.BOLD}{sequence[start:end]}{bcolors.ENDC}{sequence[end:end+retron_size_right]}\")\n",
        "        for insert in insert_pool:\n",
        "\n",
        "            retron = f\"{sequence[start-retron_size_left:start]}|{insert}|{sequence[end:end+retron_size_right]}\"\n",
        "            assert len(retron)-2 == retron_size\n",
        "            # print(f\"{sequence[start-retron_size_left:start]}{bcolors.OKGREEN}{insert}{bcolors.ENDC}{sequence[end:end+retron_size_right]}\")\n",
        "\n",
        "\n",
        "\n",
        "            fwd_primer = f'{bl}{goldl}{ovl}'\n",
        "            rev_primer = ''.join([invert_base_seq[nt] for nt in f'{ovr}{goldr}{br}'[::-1]])\n",
        "\n",
        "            retron_pool.append((insert,\n",
        "                                fwd_primer,\n",
        "                               rev_primer,\n",
        "                               f'{bl}|{goldl}|{ovl}|{retron}|{ovr}|{goldr}|{br}'))\n",
        "\n",
        "    return retron_pool\n",
        "\n",
        "# WIP\n",
        "# updated_pool = []\n",
        "# for insert in insert_pool:\n",
        "#     opti = ecoli_pref.get(codon_to_aa[insert],None)\n",
        "#     if opti is not None:\n",
        "#         updated_pool.append(opti)\n",
        "# list(set(updated_pool))\n",
        "\n",
        "from urllib.request import urlretrieve\n",
        "url_fwd = 'https://raw.githubusercontent.com/schmidt-lab/SPINE/master/SPINE/data/forward_finalprimers.fasta'\n",
        "url_rev = 'https://raw.githubusercontent.com/schmidt-lab/SPINE/master/SPINE/data/reverse_finalprimers.fasta'\n",
        "\n",
        "if not os.path.isfile('forward_finalprimers.fasta') or not os.path.isfile('reverse_finalprimers.fasta'):\n",
        "    print('Downloading barcoding lists')\n",
        "    urlretrieve(url_fwd,'forward_finalprimers.fasta')\n",
        "    urlretrieve(url_rev,'reverse_finalprimers.fasta')\n",
        "barcode_dict = {}\n",
        "with open('forward_finalprimers.fasta','r') as f:\n",
        "    for line in f:\n",
        "        if line.startswith('>'):\n",
        "            key = '-'.join(line.strip().split('-')[:2])\n",
        "        else:\n",
        "            barcode_dict[key] = line.strip()\n",
        "with open('reverse_finalprimers.fasta','r') as f:\n",
        "    for line in f:\n",
        "        if line.startswith('>'):\n",
        "            key = '-'.join(line.strip().split('-')[:2])\n",
        "        else:\n",
        "            barcode_dict[key] = [barcode_dict[key.replace('F','R')],line.strip()]\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "xzBjNCsa1aTI"
      },
      "source": [
        "## Barcode Generator"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "nYoDM5y51aTI"
      },
      "source": [
        "maximum number is 3000"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "BwBr51sI1aTI"
      },
      "outputs": [],
      "source": [
        "barcode_number = 11 # @param {type:\"integer\"}\n",
        "print(':'.join(barcode_dict[f'>skpp-{barcode_number}']))"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "I5fmtxWa1aTJ"
      },
      "source": [
        "## Systematic Screen"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "LOtJDjCHpd8G"
      },
      "outputs": [],
      "source": [
        "# @title Input parameters here:\n",
        "\n",
        "experiment_name = '' #@param {type:\"string\"}\n",
        "\n",
        "sequence = '' #@param {type:\"string\"}\n",
        "\n",
        "barcode = 'TTATAATCATCCTCCCCGGC:CCAAATAGGATGTGTGCTCG' #@param {type:\"string\"}\n",
        "goldengate_site = 'CGTCTCA:TGAGACG'#@param {type:\"string\"}\n",
        "overhang = 'ATTC:GTAC'#@param {type:\"string\"}\n",
        "start_position = 0#@param {type:\"integer\"}\n",
        "window_size = 3 #@param {type:\"integer\"}\n",
        "step_size = 3 #@param {type:\"integer\"}\n",
        "end_position = 10 #@param {type:\"raw\"}\n",
        "retron_size = 90 #@param {type:\"integer\"}\n",
        "extension = 'ATG'#@param {type:\"string\"}\n",
        "#species = 'all' #@param [\"all\",\"ecoli\"]\n",
        "\n",
        "goldengate_site = goldengate_site.split(':')\n",
        "overhang = overhang.split(':')\n",
        "barcode = barcode.split(':')\n",
        "seqs = []\n",
        "for part in extension.split(':'):\n",
        "    if special_token_mapping.get(part,None) is None:\n",
        "        seqs.extend([base_mapping[base] for base in part])\n",
        "    else:\n",
        "        seqs.append(special_token_mapping[part])\n",
        "combinations = list(product(*seqs))\n",
        "\n",
        "insert_pool = [''.join(combo) for combo in combinations]\n",
        "print(f'[INFO] - Length provided sequence: {len(sequence)} Length insert: {len(extension)}\\\n",
        " Total: {len(sequence) + len(extension)}')\n",
        "\n",
        "print('[INFO] - Inserts:',insert_pool)\n",
        "\n",
        "print(f'Barcodes: {barcode}')\n",
        "\n",
        "if end_position is None:\n",
        "  end_position = start_position\n",
        "\n",
        "overall_pool = {}\n",
        "\n",
        "for pos in range(start_position,end_position+step_size,step_size):\n",
        "  if window_size != 0:\n",
        "    target_position = [pos,pos+window_size]\n",
        "  else:\n",
        "    target_position = pos\n",
        "\n",
        "  print(f'\\nPosition - [{pos}] - POOL:\\n')\n",
        "  retronpool = generate_retrons(sequence=sequence.upper(),\n",
        "                          extension=extension,\n",
        "                          target_position=target_position,\n",
        "                          retron_size=retron_size,\n",
        "                          barcodes=barcode,\n",
        "                          goldengate_sites=goldengate_site,\n",
        "                          overhangs=overhang)\n",
        "\n",
        "\n",
        "  for insert, fwd,rev,seq in retronpool:\n",
        "    b1,gg1,ov1,left,var,right,ov2,gg2,b2 = split_seq = seq.split('|')\n",
        "    if len(insert) == 0:\n",
        "        kind = 'Del'\n",
        "    elif window_size-len(insert) == 0:\n",
        "        kind = 'Mut_'\n",
        "    else:\n",
        "        kind = 'Ins_'\n",
        "    overall_pool[f'{experiment_name}_{pos}_{kind}{insert}'] = seq.replace('|','')\n",
        "\n",
        "#     print(f\"{bcolors.OKCYAN}{b1}{bcolors.ENDC}{bcolors.OKBLUE}{gg1}{bcolors.ENDC}\\\n",
        "# {bcolors.OKGREEN}{ov1}{bcolors.ENDC}{left}{bcolors.FAIL}{var}{bcolors.ENDC}\\\n",
        "# {right}{bcolors.OKGREEN}{ov2}{bcolors.ENDC}{bcolors.OKBLUE}{gg2}{bcolors.ENDC}\\\n",
        "# {bcolors.OKCYAN}{b2}{bcolors.ENDC}\")\n",
        "#   print('\\n')\n",
        "#   print(f'FWD: {fwd}\\nREV: {rev}')\n",
        "#   print('\\n--------------------------\\n')\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4U2ZmFWRGmGT"
      },
      "source": [
        "## Saving"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "iQgCmPVT1aTJ"
      },
      "outputs": [],
      "source": [
        "with open(f'{experiment_name}_Pool.fasta','w') as f:\n",
        "    for line in gen_fasta(overall_pool):\n",
        "        f.write(line)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "alcAXCa7Gpky"
      },
      "outputs": [],
      "source": [
        "raise ValueError('Stop')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "fgdbZmKYHQcv"
      },
      "source": [
        "## Linkers at static site"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "7pBtsFgh1aTJ",
        "scrolled": false,
        "collapsed": true
      },
      "outputs": [],
      "source": [
        "# @title Input parameters here:\n",
        "\n",
        "experiment_name = '' #@param {type:\"string\"}\n",
        "\n",
        "sequence = '' #@param {type:\"string\"}\n",
        "\n",
        "barcode = 'ATATAGATGCCGTCCTAGCG:TGGGCACAGGAAAGATACTT' #@param {type:\"string\"}\n",
        "goldengate_site = 'CGTCTCA:TGAGACG'#@param {type:\"string\"}\n",
        "overhang = 'ATTC:GTAC'#@param {type:\"string\"}\n",
        "target_position = 0#@param {type:\"integer\"}\n",
        "\n",
        "retron_size = 0 #@param {type:\"integer\"}\n",
        "linker_pool = 'GG,PP,GXG,GGSG,GSGG,GPPG,GSGSG,GPPPG,GSGGSG,GPPPPG,-,--,---,+,++,+++,X'#@param {type:\"string\"}\n",
        "species = 'ecoli' #@param [\"ecoli\",\"ecoli_hydrophilic\",\"ecoli_hydrophobic\"]\n",
        "\n",
        "\n",
        "linker_pool = linker_pool.split(',')\n",
        "\n",
        "combinations = []\n",
        "for linker in linker_pool:\n",
        "    if 'X' in linker.upper():\n",
        "        combinations.extend(generate_X_variants(linker,species))\n",
        "\n",
        "    elif '-' in linker.upper() or '+' in linker.upper():\n",
        "        combinations.append(3*(linker))\n",
        "    else:\n",
        "        combinations.append(aa_to_codon_convert(linker,species))\n",
        "\n",
        "\n",
        "start_position = end_position = target_position\n",
        "window_size = 0\n",
        "\n",
        "goldengate_site = goldengate_site.split(':')\n",
        "overhang = overhang.split(':')\n",
        "barcode = barcode.split(':')\n",
        "# seqs = [base_mapping[base] for base in extension]\n",
        "# combinations = list(product(*seqs))\n",
        "\n",
        "insert_pool = [''.join(combo) for combo in combinations]\n",
        "print(f'[INFO] - Length provided sequence: {len(sequence)}\\\n",
        " Total: {len(sequence) + 3}')\n",
        "print('[INFO] - Inserts:',insert_pool)\n",
        "print(f'Barcodes: {barcode}')\n",
        "if end_position is None:\n",
        "  end_position = start_position\n",
        "\n",
        "overall_pool = {}\n",
        "pos = target_position\n",
        "if window_size != 0:\n",
        "    target_position = [pos,pos+window_size]\n",
        "else:\n",
        "    target_position = pos\n",
        "\n",
        "print(f'\\nPosition - [{pos}] - POOL:\\n')\n",
        "retronpool = []\n",
        "for extension in combinations:\n",
        "\n",
        "    if '-' in extension:\n",
        "       w = (len(extension))\n",
        "       target_position = [pos-w-window_size,pos-window_size]\n",
        "       extension = ''\n",
        "       print(target_position,window_size)\n",
        "    elif '+' in extension:\n",
        "       w = (len(extension))\n",
        "       target_position = [pos,pos+w]\n",
        "       extension = ''\n",
        "       print(target_position,window_size)\n",
        "    else:\n",
        "        target_position = pos#[pos,pos+window_size]\n",
        "        w = window_size\n",
        "    retronpool = generate_retrons(sequence=sequence.upper(),\n",
        "                          extension=extension,\n",
        "                          target_position=target_position,\n",
        "                          retron_size=retron_size,\n",
        "                          barcodes=barcode,\n",
        "                          goldengate_sites=goldengate_site,\n",
        "                          overhangs=overhang)\n",
        "\n",
        "\n",
        "    for insert, fwd,rev,seq in retronpool:\n",
        "        b1,gg1,ov1,left,var,right,ov2,gg2,b2 = split_seq = seq.split('|')\n",
        "        if len(insert) == 0:\n",
        "            kind = f'Del_{w}'\n",
        "        elif window_size-len(insert) == 0:\n",
        "            kind = 'Mut_'\n",
        "        else:\n",
        "            kind = 'Ins_'\n",
        "        overall_pool[f'{experiment_name}_{pos}_{kind}{insert}'] = seq.replace('|','')\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "FL3ivLqGIS8T"
      },
      "source": [
        "## Saving"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Y3wTsRKu1aTK"
      },
      "outputs": [],
      "source": [
        "with open(f'{experiment_name}_Pool.fasta','w') as f:\n",
        "    for line in gen_fasta(overall_pool):\n",
        "        f.write(line)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "5iZM2iIi1aTK"
      },
      "source": [
        "----"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.9.16"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}